hs_tracking/models.py

```diff
@@ -0,0 +1,147 @@
+
+
+class SessionManager(models.Manager):
+    def for_request(self, request):
+        signed_id = request.session.get('hs_tracking_id')
+        if signed_id:
+            tracking_id = signing.loads(signed_id)
+            cut_off = datetime.now() - timedelta(seconds=SESSION_TIMEOUT)
+            session = None
+            try:
+                session = Session.objects.filter(begin__gte=cut_off).get(id=tracking_id['id'])
+            except Session.DoesNotExist:
+                pass
+            if session is not None:
+                if session.visitor.user is None and request.user.is_authenticated():
+                    try:
+                        session.visitor = Visitor.objects.get(user=request.user)
+                        session.save()
+                    except Visitor.DoesNotExist:
+                        session.visitor.user = request.user
+                        session.visitor.save()
+                return session
+        # No session found, create one
+        if request.user.is_authenticated():
+            visitor, _ = Visitor.objects.get_or_create(user=request.user)
+        else:
+            visitor = Visitor.objects.create()
+        session = Session.objects.create(visitor=visitor)
+        request.session['hs_tracking_id'] = signing.dumps({'id': session.id})
+        return session
+
+
+class Visitor(models.Model):
+    first_seen = models.DateTimeField(auto_now_add=True)
+    user = models.ForeignKey(settings.AUTH_USER_MODEL, null=True, blank=True, unique=True)
+
+    def export_visitor_information(self):
+        """Exports visitor profile information."""
+        info = {
+            "id": self.id,
+        }
+        if self.user:
+            profile = UserProfile.objects.get(user=self.user)
+            for field in PROFILE_FIELDS:
+                info[field] = getattr(profile, field)
+            for field in USER_FIELDS:
+                info[field] = getattr(self.user, field)
+        else:
+            profile = None
+            for field in PROFILE_FIELDS:
+                info[field] = None
+            for field in USER_FIELDS:
+                info[field] = None
+        return info
+
+
+class Session(models.Model):
+    begin = models.DateTimeField(auto_now_add=True)
+    visitor = models.ForeignKey(Visitor)
+
+    objects = SessionManager()
+
+    def get(self, name):
+        return Variable.objects.filter(session=self, name=name).first().get_value()
+
+    def getlist(self, name):
+        return [v.get_value() for v in Variable.objects.filter(session=self, name=name)]
+
+    def record(self, *args, **kwargs):
+        args = (self,) + args
+        return Variable.record(*args, **kwargs)
+
+
+class Variable(models.Model):
+    TYPES = (
+        ('Integer', int),
+        ('Floating Point', float),
+        ('Text', unicode),
+        ('Flag', bool),
+    )
+    TYPE_CHOICES = (
+        (i, label)
+        for (i, label) in
+        enumerate((label for (label, coercer) in TYPES), 1)
```

Comment:
@ironfroggy OK, I misunderstood the start parameter and I think I get it now.  I guess you just want the enumerator to start at 1 (rather than the default 0) for better readibility, in any case, I don't see an issue with starting the enumerator at 1 now.